dojo/pipeline.py

```diff
@@ -69,6 +75,63 @@ def modify_permissions(backend, uid, user=None, social=None, *args, **kwargs):
+        soc = user.social_auth.get()
+        token = soc.extra_data['access_token']
+        group_names = []
+        if 'groups' not in kwargs['response'] or kwargs['response']['groups'] == "":
+            logger.warn("No groups in response. Stopping to update groups of user based on azureAD")
+            return
+        group_IDs = kwargs['response']['groups']
+        for group_from_response in group_IDs:
+            logger.debug("Analysing Group_ID " + group_from_response)
+            request_headers = {'Authorization': 'Bearer ' + token}
+            try:
+                if is_group_id(group_from_response):
+                    logger.debug("detected " + group_from_response + " as groupID and will fetch the displayName from microsoft graph")
+                    group_name_request = requests.get((str(soc.extra_data['resource']) + '/v1.0/groups/' + str(group_from_response) + '?$select=displayName'), headers=request_headers)
+                    group_name_request_json = group_name_request.json()
+                    group_name = group_name_request_json['displayName']
+                else:
+                    logger.debug("detected " + group_from_response + " as group name and will not call microsoft graph")
+                    group_name = group_from_response
+
+                group, created_group = Dojo_Group.objects.get_or_create(name=group_name, is_azure=True)
+                if settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER == "" or re.search(settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER, group_name):
+                    group_names.append(group_name)
+                else:
+                    logger.debug("Skipping group " + group_name + " due to AZUREAD_TENANT_OAUTH2_GROUPS_FILTER " + settings.AZUREAD_TENANT_OAUTH2_GROUPS_FILTER)
+                    continue
+
+                group_member, is_member_created = Dojo_Group_Member.objects.get_or_create(group=group, user=user, defaults={
+                    'role': Role.objects.get(id=Roles.Writer)})
+
+                if settings.AZUREAD_TENANT_OAUTH2_CLEANUP_GROUPS:
+                    cleanup_old_azureAD_groups_for_user(user, group_names)
+            except:
+                logger.error("Could not call microsoft graph API or save groups to member")
+                traceback.print_exc()
+
+
+def is_group_id(group):
+    if re.search(r'^[a-zA-Z0-9]{8,}-[a-zA-Z0-9]{4,}-[a-zA-Z0-9]{4,}-[a-zA-Z0-9]{4,}-[a-zA-Z0-9]{12,}$', group):
+        return True
+    else:
+        return False
+
+
+def cleanup_old_azureAD_groups_for_user(user, group_names):
+    all_groups = Dojo_Group.objects.all()
+    for group in all_groups:
+        member = Dojo_Group_Member.objects.filter(group=group, user=user)  # A user shouldn't have multiple roles (see code above) and therefore only one member will be returned
+        logger.debug("Group to delete: " + str(group))
+        if member:
+            if str(group) not in group_names and group.is_azure:
```

Comment:
slightly better would be to include `group__is_azure==True` to the `Dojo_Group_Member` filter query.

Reply:
what if there is a group "SecurityAnalysts" that was created manually in DD, so is_azure is false. Then a AAD users signs in which has that role in its token. The user is added to the group "SecurityAnalysts". Then the user is removed from that group in AAD. Then on the next login the user will not be removed from the group in DD because it is not an azure group. Not sure if this the expected behaviour.

Reply:
I am not sure how to make it better. Shall we document that case under the cleanup flag?

Reply:
Maybe. Currently there is no way for users/admins to see which groups are "local groups" and which groups are "azure groups". Safest would be to remove the user from all types of groups to avoid permission creep?

Reply:
@valentijnscholten Do you recommend to keep the azuread flag in case we remove the user membership to the group in any case or should we remove it?

Reply:
@valentijnscholten ?

Reply:
it might be useful to know where the group came from, but then it should be visible somewhere probably

Reply:
I have removed the check for azuread but I keep azuread for future options. Is it ok now?

Reply:
ping @valentijnscholten

Reply:
@Maffooch could you perform a review?